module/VuFind/src/VuFind/Search/SearchOrigin/SearchOriginFactory.php

```diff
@@ -0,0 +1,72 @@
+<?php
+
+/**
+ * Search Origin Object
+ *
+ * PHP version 8
+ *
+ * Copyright (C) Villanova University 2010.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * @category VuFind
+ * @package  TODO?
+ * @author   Robby ROUDON <roudonro@msu.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Page
+ */
+
+namespace VuFind\Search\SearchOrigin;
+
+use Exception;
+
+/**
+ * Factory for search origin objects
+ *
+ * @category VuFind
+ * @package  TODO?
+ * @author   Robby ROUDON <roudonro@msu.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Page
+ */
+class SearchOriginFactory
+{
+    /**
+     * From a request return an AbstractSearchOrigin object
+     *
+     * @param array $params URL GET parameters
+     *
+     * @return AbstractSearchOrigin|null
+     */
+    public function createObject(array $params): ?AbstractSearchOrigin
+    {
+        $nameParam = AbstractSearchOrigin::PARAM_NAME;
+        if (empty($params[$nameParam])) {
+            return null;
+        }
+        try {
+            return match ($params[$nameParam]) {
```

Comment:
It might make sense to refactor this match to a protected support method, in order to make it easier to register custom origins in a subclass without having to copy and paste the entire method.

Reply:
I don't understand what you mean

Reply:
I'm just suggesting refactoring the contents of the `try` here to a separate support method like:

```
        protected function createObjectByName(string $name)
        {
            return match ($name) {
                AlphaBrowseSearchOrigin::getName() => new AlphaBrowseSearchOrigin(
                    $params[AlphaBrowseSearchOrigin::SEARCH_SOURCE_DISPLAY_PARAM] ?? null,
                    $params[AlphaBrowseSearchOrigin::SEARCH_SOURCE_PARAM] ?? null,
                    $params[AlphaBrowseSearchOrigin::SEARCH_FROM_PARAM] ?? null,
                    $params[AlphaBrowseSearchOrigin::SEARCH_PAGE_PARAM] ?? null
                ),
                default => null,
            };
        }
```

This would make it easier for somebody to build a custom factory, since they could add a subclass method like:

```
        protected function createObjectByName(string $name)
        {
            return $name === 'myCustomThing' ? $this->buildMyCustomObject() : parent::createObjectByName($name);
        }
```

Maybe this type of customization is unlikely to be necessary and it's not worth making the change... but it seemed like it could be beneficial in the future if we offered this possibility.

Reply:
In the way it currently is, the try only catches the exception thrown when the search origin name is matched but not all the parameters are present, which in case of an extension won't pose any problem, unless I didn't understand correctly

Reply:
Yes, I'm not suggesting that we should change the exception handling or the way this is called externally. I'm just suggesting, in combination with my new method proposed above, the main createObject could become:

```
    public function createObject(array $params): ?AbstractSearchOrigin
    {
        $nameParam = AbstractSearchOrigin::PARAM_NAME;
        if (empty($params[$nameParam])) {
            return null;
        }
        try {
            return $this->createObjectByName($params[$nameParam]);
        } catch (Exception $e) {
            $this->logWarning('Error while trying to build search origin object : ' . $e->getMessage());
            return null;
        }
    }
```

That just separates the input validation and exception handling logic from the actual object building logic, which helps with extensibility (and potentially readability, especially if this grows in future).